### PR TITLE
Refactor MonteCarloSimulation class

### PR DESCRIPTION
**Pull request by mlund**
# Description

This refactors `MCSimulations` into `MonteCarloSimulation` with focus on readability and preparation for future changes such as external builders as an alternative to current JSON based constructors.

### CODE REVIEW

File: src/montecarlo.cpp [hunk 1]
```diff
@@ -1,231 +1,259 @@
...
+                double du = unew - uold;                         // potential energy change (kT)
+                if (std::isnan(uold) and not std::isnan(unew)) { // if NaN --> finite energy change
+                    du = pc::neg_infty;                          // ...always accept
+                } else if (std::isnan(unew)) {                   // if moving to NaN, e.g. division by zero,
+                    du = pc::infty;                              // ...always reject
+                } else if (std::isnan(du)) {                     // if difference is NaN, e.g. infinity - infinity,
+                    du = 0.0;                                    // ...always accept
                 }
+                double bias = (*move)->bias(change, uold, unew); // moves *may* add bias (kT)
+                double ideal = TranslationalEntropy(*new_state->spc, *old_state->spc).energy(change);
```
Comment (rc83) [inline src/montecarlo.cpp:133]: Explicitly comment on that this is only relevant for the speciation move. Is `ideal` really descriptive here?

File: src/montecarlo.cpp [hunk 2]
```diff
@@ -1,231 +1,259 @@
...
-                } else { // reject move
-                    state2.sync(state1, change);
-                    (**mv).reject(change);
-                    du = 0;
-                }
-                dusum += du; // sum of all energy changes
-            }
+/**
+ * @param N_new Number of atoms or molecules after move
+ * @param N_old Number of atoms or molecular before move
```
Comment (rc83) [inline src/montecarlo.cpp:180]: Variable names with a capital letter does not fit into our naming convention.

File: src/montecarlo.h [hunk 3]
```diff
@@ -2,82 +2,121 @@
...
+namespace MPI {
+class MPIController;
+}
 
+/**
+ * @brief Class to handle Monte Carlo moves
+ *
+ * This class maintains a list of MC moves and takes care of
+ * randomly selecting a moves, perform the move, decide if the move
+ * if accepted, and also checks if there is a drift in the potential
```
Comment (rc83) [inline src/montecarlo.h:28]: if → is

File: src/montecarlo.h [hunk 4]
```diff
@@ -2,82 +2,121 @@
...
+ * energy.
+ *
+ * The moves always operate on a trial state (`new_state`). If accepted,
+ * the changes will be copied into the `old_state`. The state here refers
```
Comment (rc83) [inline src/montecarlo.h:32]: Why not to name the state variables more descriptively, e.g., `trial_state` and `default_state` or similarly?

File: src/montecarlo.h [hunk 5]
```diff
@@ -2,82 +2,121 @@
...
-    */
-    void restore(const json &j); //!< restore system from previously store json object
-    void move();
-    void to_json(json &j);
+    MetropolisMonteCarlo(const json &, MPI::MPIController &);
+    Energy::Hamiltonian &getHamiltonian();                     //!< Get Hamiltonian of accepted (old) state
+    Space &getSpace();                                         //!< Access to space in accepted (old) state
+    double relativeEnergyDrift();                              //!< Relative energy drift from initial configuration
+    void move();                                               //!< Perform random Monte Carlo move
+    void restore(const json &);                                //!< restore system from previously store json object
```
Comment (rc83) [inline src/montecarlo.h:79]: Comment should start with a capital letter. ;)

File: src/group.h [hunk 6]
```diff
@@ -44,7 +44,9 @@ namespace Faunus {
     template<class T>
         class ElasticRange : public IterRange<typename std::vector<T>::iterator> {
             public:
-                typedef typename std::vector<T>::iterator Titer;
+              using Titer = typename std::vector<T>::iterator;
+              using const_iterator = typename std::vector<T>::const_iterator;
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/group.h:48]: Yes, I prefer the lower case like in STL, but didn't want to clutter the PR with a lot of changes in group.h. Any thoughts?

File: src/montecarlo.h (hunk 4 above)
Comment (mlund) [reply src/montecarlo.h:32]: indeed, `trial_state` and merely `state`?

File: src/group.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (mlund) [reply src/group.h:48]: Will leave it as is and make a separate commit for group as several things could be improved there.

**Pull request merged**